maven-core/src/main/java/org/apache/maven/execution/DefaultBuildResumptionAnalyzer.java

```diff
@@ -47,132 +41,49 @@
-    }
-
-    /**
-     * Projects after the first failed project could have succeeded by using -T or --fail-at-end.
-     * These projects can be skipped from later builds.
-     * This is not the case these projects are dependent on one of the failed projects.
-     * @param result The result of the Maven build.
-     * @param failedProjects The list of failed projects in the build.
-     * @param startFromProjectIndex Start looking for projects which can be skipped from a certain index.
-     * @return A list of projects which can be skipped in a later build.
-     */
-    private List<String> determineProjectsToSkip( MavenExecutionResult result,
-                                                  List<MavenProject> failedProjects,
-                                                  int startFromProjectIndex )
-    {
-        List<MavenProject> allProjects = result.getTopologicallySortedProjects();
-        List<MavenProject> remainingProjects = allProjects.subList( startFromProjectIndex, allProjects.size() );
-
-        List<GroupArtifactPair> failedProjectsGAList = failedProjects.stream()
-                .map( GroupArtifactPair::new )
-                .collect( Collectors.toList() );
-
-        return remainingProjects.stream()
+        long succeeded = sortedProjects.stream()
                 .filter( project -> result.getBuildSummary( project ) instanceof BuildSuccess )
-                .filter( project -> hasNoDependencyOnProjects( project, failedProjectsGAList ) )
-                .map( project -> project.getGroupId() + ":" + project.getArtifactId() )
-                .collect( Collectors.toList() );
-    }
-
-    private boolean hasNoDependencyOnProjects( MavenProject project, List<GroupArtifactPair> projectsGAs )
-    {
-        return project.getDependencies().stream()
-                .map( GroupArtifactPair::new )
-                .noneMatch( projectsGAs::contains );
-    }
-
-    private static class GroupArtifactPair
-    {
-        private final String groupId;
-        private final String artifactId;
-
-        GroupArtifactPair( MavenProject project )
-        {
-            this.groupId = project.getGroupId();
-            this.artifactId = project.getArtifactId();
-        }
+                .count();
 
-        GroupArtifactPair( Dependency dependency )
+        if ( succeeded == 0 )
         {
-            this.groupId = dependency.getGroupId();
-            this.artifactId = dependency.getArtifactId();
+            return Optional.empty();
         }
 
-        @Override
-        public boolean equals( Object o )
+        List<MavenProject> projects = sortedProjects.stream()
+                .filter( project -> result.getBuildSummary( project ) == null
+                        || result.getBuildSummary( project ) instanceof BuildFailure )
+                .collect( Collectors.toList() );
+        while ( true )
         {
-            if ( this == o )
+            List<MavenProject> children = projects.stream()
```

Comment:
I haven't checked the possible impact of `--non-recursive`.  My main motivation was to keep the list small, especially as the `--resume` will now behave as `--pl` and that one automatically adds children projects to the build, see https://github.com/apache/maven/blob/22016982341262193d802ef51b2c236bac98464a/maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java#L194-L198
I don't think this is impacted by the `--non-recursive`.  What could be the behavior when using both `--resume` and `--non-recursive` ? Not sure it makes much sense.

Reply:
As far as I remember `--non-recursive` will result in `getCollectedProjects` to be empty (even if the pom has child modules). Actually --non-recursive will _only_ read the target pom (either ./pom.xml or the pom file selected via `-f pom.xml`, and none of it's child projects, I guess). So any of the selected projects will not take effect anyway.

My preference would go out to listing all modules in the BuildResumptionData regardless, because:
- the behavior in `DefaultGraphBuilder` could change,
- users will not typically open `resume.properties`, so we do not have to optimize the flag for readability,
- this `while(true)` loop adds complexity to the otherwise concise method.

Reply:
Just checked the combination of `-r` and `--non-recursive`, and it doesn't work:

> [INFO] Scanning for projects...
> [INFO] Resuming from org.apache.maven:maven-plugin-api, org.apache.maven:maven-xml, org.apache.maven:maven-model-builder, org.apache.maven:maven-settings-builder, org.apache.maven:maven-resolver-provider, org.apache.maven:maven-core, org.apache.maven:maven-slf4j-provider, org.apache.maven:maven-embedder, org.apache.maven:maven-compat, org.apache.maven:apache-maven, org.apache.maven:maven-wrapper, org.apache.maven:apache-maven-wrapper due to the --resume / -r feature.
> [ERROR] [ERROR] Could not find the selected project in the reactor: org.apache.maven:maven-plugin-api

This makes sense as the modules listed in **target/resume.properties** do not exist in the Reactor. `--non-recursive` effectively reduces the available projects to only the selected project, as @MartinKanters pointed out.

It's not something that's changing with these code changes, so let's keep it this way and see if the message is clear enough for end-users.